### PR TITLE
Initial mcts implementation, WIP

### PR DESCRIPTION
**Pull request by PhilipEkfeldt**
Initial code for #5 , any feedback appreciated. Not ready for merge

### CODE REVIEW

**Comment by charlesoblack** [comment]
I'd like to move the definition of `Game` to a different file, if that's alright. I'm currently writing it up on a `_game.py` file. I'll do a PR soon so you can take a look.

**Comment by PhilipEkfeldt** [comment]
> I'd like to move the definition of `Game` to a different file, if that's alright. I'm currently writing it up on a `_game.py` file. I'll do a PR soon so you can take a look.

Of course, I just did it this way for removing linting warning :p

**Comment by PhilipEkfeldt** [comment]
@charlesoblack please review. This is an initial implementation to close #5, I will create a new issue with improvements that are needed.

File: alphazero/__init__.py [hunk 1]
```diff
@@ -1,4 +1,3 @@
 #! /usr/bin/python3
-
```
Comment (charlesoblack) [inline alphazero/__init__.py:2]: pep 8 my man... :( the newline has to be there!

File: alphazero/_game.py [hunk 2]
```diff
@@ -195,9 +195,10 @@ def board_after_move(board, player: int, move: (int, int)):
                 f"invalid move {move}: space already occupied in board {board}"
             )
 
-        board[move] = player
+        new_board = board.clone()
+        new_board[move] = player
 
-        return board
+        return new_board
```
Comment (charlesoblack) [inline alphazero/_game.py:201]: is there a reason for returning a copy of the board instead of the in-place change? I'm guessing this is somehow interfering with your MCTS. If so that's fine, I just want to make sure :)

File: alphazero/_game.py [hunk 3]
```diff
@@ -228,11 +229,11 @@ def current_legal_moves(self) -> torch.Tensor:
...
     def get_game_status(board) -> tuple:
-        verts_0 = any((board == 0).sum(axis=0) == 3)
-        horizontals_0 = any((board == 0).sum(axis=1) == 3)
+        verts_0 = any((board == 0).sum(dim=0) == 3)
+        horizontals_0 = any((board == 0).sum(dim=1) == 3)
 
-        verts_1 = any((board == 1).sum(axis=0) == 3)
-        horizontals_1 = any((board == 1).sum(axis=1) == 3)
+        verts_1 = any((board == 1).sum(dim=0) == 3)
+        horizontals_1 = any((board == 1).sum(dim=1) == 3)
```
Comment (charlesoblack) [inline alphazero/_game.py:236]: is this one of those deprecated kwargs things that pytorch didn't warn me about?

File: alphazero/_game.py [hunk 4]
```diff
@@ -262,10 +263,7 @@ def is_game_over(board) -> bool:
         just return a boolean True/False.
         """
 
-        if any(TicTacToe.get_game_status(board)):
-            return True
-        # implicit else
-        return False
+        return not TicTacToe.result(board) is None
```
Comment (charlesoblack) [inline alphazero/_game.py:266]: definitely feel like this line could be clarified a lot more... maybe capture the result separately, then eval if none, and then return? I find this line hard to read

File: alphazero/mcts/mcts.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+#    game: Game,
+#    net: torch.nn.Module,
+#    n_mcts_iter: int = 1600,
+#    n_games_thread=2500,
+#    threads=4,
+# ):
+#    # Arg games/thread instead?
+#    def play_games_thread():
+#        data = []
+#        for i in range(n_games_thread):
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:282]: can we remove this? maybe you can add it to your stash to keep it around?

File: alphazero/utils/_utils.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import numpy as np
+
+
+def sample_tensor_indices(tensor, n_samples):
```
Comment (charlesoblack) [inline alphazero/utils/_utils.py:4]: can this get a docstring please? :P

File: alphazero/utils/_utils.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+    ind_1d = tensor.float().flatten().multinomial(n_samples)
+    samples = np.unravel_index(ind_1d, tensor.size())
+    samples = np.array(samples)
+    samples = [tuple(r) for r in samples.T]
```
Comment (charlesoblack) [inline alphazero/utils/_utils.py:8]: do we need the transpose here?

File: tests/test_game.py [hunk 8]
```diff
@@ -284,3 +284,8 @@ def test_json_accepts_any_order_of_params(self):
         ttt_json = TicTacToe.from_json(**json_dict)
 
         assert torch.all(ttt.board_state.eq(ttt_json.board_state))
+
+    def test_tie_board_is_game_over(self):
+        board = torch.Tensor([[1, 0, 1], [1, 0, 1], [0, 1, 0]])
+
+        assert TicTacToe.is_game_over(board)
```
Comment (charlesoblack) [inline tests/test_game.py:291]: can we clarify that this assert should be true? I find it more readable (despite being crappy efficiency practice)

File: tests/test_mcts.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import pytest
+from alphazero.mcts import mcts, MCTSNode
+from alphazero import TicTacToe
+import torch
+
+
+class TestMCTS:
+    def test_tictactoe_first_step(self):
+        policy, _ = mcts(
```
Comment (charlesoblack) [inline tests/test_mcts.py:9]: do we wanna check the Q value here too? or is this going to give us a random value that can't be tested?

File: tests/test_utils.py [hunk 10]
```diff
@@ -1,8 +1,16 @@
 #! /usr/bin/python3
-
```
Comment (charlesoblack) [inline tests/test_utils.py:2]: pep8 :(

File: alphazero/mcts/mcts.py [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+class TicTacToeConverter(BoardConverter):
+    @staticmethod
+    def board_to_tensor(board):
+        player_1 = board == 0
+        player_2 = board == 1
+        player = 1 if player_1.sum() > player_2.sum() else 0
+        player_layer = torch.full(board.size(), player)
+        return torch.stack((player_1, player_2, player))
+
+    ## TODO: Flesh this out and move
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:35]: guessing this means moving this to a different file? Since we're eventually going to use it for the network, maybe it should go under utils?

File: alphazero/mcts/mcts.py [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+import torch
+from alphazero import TicTacToe
+from alphazero.utils import sample_tensor_indices
+
+GameClassType = TypeVar("T", bound="Game")
+
+
+class BoardConverter:
+    @staticmethod
+    def board_to_tensor(board):
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:13]: can you also add docstrings to all of the functions in this file? :x

File: alphazero/mcts/mcts.py [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+            game (GameClassType): Game that is currently played
+
+        Returns:
+            result (int): Final result of rollout
+        """
+        state = self._state
+        while not game.is_game_over(state):
+            legal_actions = game.get_legal_moves(state)
+            action = sample_tensor_indices(legal_actions, 1)[0]
+            player = 1 if (state == 0).sum() > (state == 1).sum() else 0
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:84]: I think we should probably move this (and the corresponding line above) to its own function in Game, or something like that. I don't like the repetition

File: alphazero/mcts/mcts.py [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+            state = game.board_after_move(state, player, action)
+        result = game.result(state)
+        return result
+
+    def calc_policy_value(
+        self, rollout: bool, game: GameClassType, network: torch.nn.Module = None
+    ) -> (torch.Tensor, float):
+        """
+        Calculates the policy and value for the node. If rollout performs a rollout, else uses network to get policy and value
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:93]: Can you separate this into two lines?

(this is one of those things that black doesn't do for you... I think we should keep to a 80 char limit (or so) per line. It helps me read code with my tiny 768p screen :P )

File: alphazero/mcts/mcts.py [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+            game (GameClassType): Game that is played
+            network (torch.nn.Module, optional): Optional, must be set if rollout is false
+
+        Returns:
+            policy (torch.Tensor): The calculated policy
+            value (float): The calculated value
+        """
+        if not rollout:
+            # TODO: Normalize probabilities after filter legal actions
+            self._policy, self._value = network(self._state)
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:105]: Since you mention in line 97 that network must be set, maybe catch the eventual error that will happen here when someone calls it without setting the network kwarg?

File: alphazero/mcts/mcts.py [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+        else:
+            self._value = self.rollout(game)
+            policy = game.get_legal_moves(self._state)
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:108]: just to be clear - when we do rollout, we are still sampling uniformly from the legal moves to get the policy? I was under the impression you could have a rollout even using a network policy.

File: alphazero/mcts/mcts.py [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+        return self._policy, self._value
+
+    def expand(self, game: GameClassType):
+        """
+        Expands node
+        Args:
+            game (GameClassType): Game that is played
+        """
+        legal_actions = game.get_legal_moves(self._state)
+        for index in legal_actions.nonzero():
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:119]: are we sure this works with False? I have no idea what goes on internally in pytorch so I don't want to assume.

File: alphazero/mcts/mcts.py [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+            action = torch.zeros(legal_actions.size())
+            action[tuple(index)] = 1
+            prior = self._policy[tuple(index)].item()
+            player = 1 if (self._state == 0).sum() > (self._state == 1).sum() else 0
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:123]: same as above.

File: alphazero/mcts/mcts.py [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+
+    @property
+    def children(self):
+        return self._children
+
+    @property
+    def is_root(self):
+        return self.parent is None
+
+    def u_value(self, sum_n: int, c_puct: float = 1.0):
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:158]: can you point me to where this is defined? I'm guessing it's in the alphagozero paper somewhere?

File: alphazero/mcts/mcts.py [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+        return c_puct * self.prior * np.sqrt(sum_n) / (1 + self.n_visit)
+
+    def confidence_bound(self, sum_n: int, c_puct: float = 1.0):
+        return self.u_value(sum_n, c_puct) + self.q_value
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:162]: same as 158

File: alphazero/mcts/mcts.py [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+        Runs a forward pass through the tree and then backpropagates back through the tree
+        Args:
+            root_node (MCTSNode): Root node to start from
+        """
+        current = root_node
+        while not current.is_leaf and not current.is_terminal:
+            sum_n = np.sum([child.n_visit for child in current.children])
+            current = current.children[
+                np.argmax(
+                    [
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:225]: for real, black?

File: alphazero/mcts/mcts.py [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+    )
+
+
+def self_play(
+    game: GameClassType, net: torch.nn.Module, n_mcts_iter: int = 1600,
+) -> List[tuple]:
+    pos = game.get_init_board()
+    temperature = 0.0001
+    data = []
+    is_terminal, res = game.is_terminal(pos)
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:253]: I think this has to be changed to reflect how Game works now

File: alphazero/mcts/mcts.py [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+        policy, node = mcts(
+            start_node=node,
+            game=game,
+            net=net,
+            n_iter=n_mcts_iter,
+            temperature=temperature,
+        )
+        data.append((pos, policy))
+        pos = game.make_move(pos, node.action)
+        is_terminal, res = game.is_terminal(pos)
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:265]: same as 253

File: alphazero/mcts/mcts.py [hunk 24]
```diff
@@ -0,0 +1,282 @@
...
+        # TODO: Currently does not add last state (when game ends, no moves available) to datapoints, should it? Also look into stopping thresholds and resigning
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:266]: can we have this in two lines too?

File: alphazero/mcts/mcts.py [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+    while not is_terminal:
+        policy, node = mcts(
+            start_node=node,
+            game=game,
+            net=net,
+            n_iter=n_mcts_iter,
+            temperature=temperature,
+        )
+        data.append((pos, policy))
+        pos = game.make_move(pos, node.action)
```
Comment (charlesoblack) [inline alphazero/mcts/mcts.py:264]: same as 253

File: alphazero/_game.py (hunk 2 above)
Comment (PhilipEkfeldt) [reply alphazero/_game.py:201]: yeah it's mutating the state I put in which is not what I want :P

File: alphazero/_game.py (hunk 3 above)
Comment (PhilipEkfeldt) [reply alphazero/_game.py:236]: Could be, torch generally uses dim instead of axis, at least in newer versions

File: alphazero/mcts/mcts.py (hunk 5 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:282]: Sure, I was just lazy

File: alphazero/utils/_utils.py (hunk 7 above)
Comment (PhilipEkfeldt) [reply alphazero/utils/_utils.py:8]: To get the correct dimension it's needed. Can flip it some other way but will be more convoluted

File: tests/test_mcts.py (hunk 9 above)
Comment (PhilipEkfeldt) [reply tests/test_mcts.py:9]: Sure we could, good point

File: alphazero/mcts/mcts.py (hunk 13 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:84]: Yup, this was just a temp fix for tictactoe. Either the mcts knows number of players and just rotates, or this is somehow returned by game. Either works.

File: alphazero/mcts/mcts.py (hunk 14 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:93]: Haha yeah, I actually got a linting warning but ignored it, my bad

File: alphazero/mcts/mcts.py (hunk 17 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:119]: It seems to work with byte tensors, so think so?

File: alphazero/mcts/mcts.py (hunk 21 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:225]: What can I say...

File: alphazero/mcts/mcts.py (hunk 16 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:108]: From what I understand, network policy replaces rollout. I guess you could get the policy from a network, but the value from a rollout? And you use the policy from the network to get the sampling probability. But I was imagining the rollout version to just be for benchmarking against the network model.

File: alphazero/mcts/mcts.py (hunk 15 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:105]: Good point

File: tests/test_utils.py (hunk 10 above)
Comment (PhilipEkfeldt) [reply tests/test_utils.py:2]: :(

File: alphazero/utils/_utils.py (hunk 6 above)
Comment (PhilipEkfeldt) [reply alphazero/utils/_utils.py:4]: sure

File: alphazero/mcts/mcts.py (hunk 22 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:253]: Yup. I'll either just stash non-used code for now or updated it

**Comment by PhilipEkfeldt** [comment]
I'll clean up based on comments and stash some unused code as well

File: alphazero/_game.py (hunk 2 above)
Comment (charlesoblack) [reply alphazero/_game.py:201]: Cool, just making sure.

**Comment by charlesoblack** [comment]
Take your time :) and thanks again for this!

File: alphazero/mcts/mcts.py (hunk 13 above)
Comment (charlesoblack) [reply alphazero/mcts/mcts.py:84]: we have access to n_players now, fyi :)

File: alphazero/mcts/mcts.py (hunk 16 above)
Comment (charlesoblack) [reply alphazero/mcts/mcts.py:108]: hmm, alright. maybe we should clarify on call next time we have one.

File: alphazero/mcts/mcts.py (hunk 17 above)
Comment (charlesoblack) [reply alphazero/mcts/mcts.py:119]: we should probably have a test for this just in case then

File: alphazero/utils/_utils.py (hunk 7 above)
Comment (charlesoblack) [reply alphazero/utils/_utils.py:8]: oh gotcha. I was just unsure what the dims of `samples` is, I thought if you iterated over it you'd just get each sample (and not the list of all samples like I'm guessing what you'd get without the transpose)

File: alphazero/_game.py (hunk 4 above)
Comment (PhilipEkfeldt) [reply alphazero/_game.py:266]: Not sure what you mean, I'll leave that for you to change

File: tests/test_mcts.py (hunk 9 above)
Comment (PhilipEkfeldt) [reply tests/test_mcts.py:9]: I think this requires too much unnecessary refactoring to get that information out so I'm going to wait with doing this. Remind me to create an issue for this

File: alphazero/mcts/mcts.py (hunk 11 above)
Comment (PhilipEkfeldt) [reply alphazero/mcts/mcts.py:35]: Stashed for now

File: alphazero/mcts/mcts.py (hunk 17 above)
Comment (charlesoblack) [reply alphazero/mcts/mcts.py:120]: image
